### PR TITLE
feat(info): implement tag management endpoints (GET/POST /info/tags)

### PR DESCRIPTION
**Pull request by SudipSinha**
## Summary
- Implement `GET /info/tags` — returns tag counts per model (optional `modelId` query param for single model or all models)
- Implement `POST /info/tags` — apply tags retroactively to row ranges with validation and idempotency
- Fix tag constant mismatch bug in `consumer_endpoint.py` (local constants shadowed `constants.py`)
- Add dual-matching for legacy unprefixed tags in `data_source.py`
- 17 tests covering happy paths and edge cases

## Bug fix included
The consumer endpoint defined local `SYNTHETIC_TAG = "synthetic"` and `UNLABELED_TAG = "unlabeled"` that shadowed the prefixed constants from `constants.py`. This caused `get_organic_dataframe()` to never match ingested data. Fixed by importing canonical constants.

## Test plan
- [x] GET /info/tags returns correct tag counts for a model
- [x] GET /info/tags with no modelId returns all models
- [x] POST /info/tags applies tags and verifies with GET
- [x] POST rejects `_trustyai` prefix (400)
- [x] POST rejects out-of-bounds indices (400)
- [x] Idempotency: applying same tag twice doesn't duplicate
- [x] Edge cases: negative indices, overlapping ranges, singleton ranges, empty dataTagging
- [x] CI: tests pass on Python 3.12 and 3.14
- [x] CI: lint, type-check, bandit all pass

Ref: RHOAIENG-55914

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added full tag management support, including viewing existing tags and applying tags to selected rows.
  * Added the ability to retrieve rows filtered by a specific tag.

* **Bug Fixes**
  * Improved handling of synthetic rows so they’re filtered consistently.
  * Added better validation for tag ranges and row bounds, with clearer error handling.

* **Tests**
  * Added coverage for tag counts, tag application, invalid inputs, and row-filtering behavior.

### CODE REVIEW

**Comment by SudipSinha** [comment]
The CodeQL `py/path-injection` alert on `pvc.py:374` is a pre-existing issue (not introduced by this PR) — it's surfaced because this PR adds a new code path (`_persist_metadata` → `delete_dataset`) that reaches the same sink.

Fixed in #240 (path traversal validation in `allocate_valid_dataset_name`).

File: src/trustyai_service/endpoints/metadata.py [hunk 1]
```diff
@@ -461,20 +465,220 @@ async def remove_column_names(request: ModelIdRequest) -> dict[str, str]:
...
+) -> None:
+    """Delete-and-rewrite the metadata dataset for a model.
+
+    .. todo:: Read-modify-write is not atomic; concurrent tag requests may conflict.
+    """
+    metadata_dataset = model_id + METADATA_SUFFIX
+    try:
+        await storage_interface.delete_dataset(metadata_dataset)
+        await storage_interface.write_data(metadata_dataset, metadata, metadata_names)
+    except Exception as exc:
```
[the remark this answers is not in the document]
Comment (SudipSinha) [reply src/trustyai_service/endpoints/metadata.py:824]: Addressed in d7c72d9. `_persist_metadata` now reads and backs up the existing data before delete. If `write_data()` fails after `delete_dataset()`, the original data is restored (best-effort).

The underlying path injection concern (user-controlled `model_id` reaching filesystem paths) is fixed separately in #240.

File: src/trustyai_service/service/data/datasources/data_source.py [hunk 2]
```diff
@@ -168,12 +170,83 @@ async def get_organic_dataframe(
...
-        if UNLABELED_TAG in df.columns:
-            df = df[~df[UNLABELED_TAG].fillna(value=False)]
+        # Filter out synthetic rows — match both the current prefixed tag and
+        # the legacy unprefixed value for backward compatibility with data
+        # written before the tag-prefix migration.
+        for col in (SYNTHETIC_TAG, "synthetic"):
+            if col in df.columns:
+                df = df[~df[col].fillna(value=False)]
 
         return df
```
[the remark this answers is not in the document]
Comment (RobGeada) [reply src/trustyai_service/service/data/datasources/data_source.py]: Yep, I agree

[the remark this answers is not in the document]
Comment (RobGeada) [reply src/trustyai_service/service/data/datasources/data_source.py]: Yeah, using columns here might be a bug in the original impl too

[the remark this answers is not in the document]
Comment (SudipSinha) [reply src/trustyai_service/service/data/datasources/data_source.py]: Addressed in c157dfce52f3763315563f39530c14d26fad16fc

[the remark this answers is not in the document]
Comment (SudipSinha) [reply src/trustyai_service/service/data/datasources/data_source.py]: Fixed in a371522 — now checks both `SYNTHETIC_TAG` (`"_trustyai_synthetic"`) and legacy `"synthetic"` when filtering the tags list.

File: src/trustyai_service/endpoints/metadata.py [hunk 3]
```diff
@@ -468,20 +476,276 @@ async def remove_column_names(
...
+                logger.exception(
+                    "Write failed for model=%s AND backup restore failed",
+                    model_id,
+                )
+
+        logger.exception("Error persisting tagged metadata for model=%s", model_id)
+        raise HTTPException(
+            status_code=HTTPStatus.INTERNAL_SERVER_ERROR,
+            detail=f"Error persisting tags for model={model_id}",
+        ) from exc
```
Comment (RobGeada) [inline src/trustyai_service/endpoints/metadata.py:845]: We shouldn't expose unmoderated exceptions back to the user - we should adopt a similar "check server logs for info" kind of message here, imo - we should swap the logger.exception and HTTPException messages

File: src/trustyai_service/service/data/datasources/data_source.py [hunk 4]
```diff
@@ -181,12 +183,94 @@ async def get_organic_dataframe(
...
+            return pd.DataFrame(df_data)
+
+        except Exception as e:  # Broad catch intentional: dataframe creation involves dynamic storage operations
+            logger.exception(
+                "Error creating dataframe by tag for model=%s, tag=%s",
+                model_id,
+                tag,
+            )
+            msg = f"Error creating dataframe by tag for model={model_id}: {e!s}"
+            raise DataframeCreateError(msg) from e
```
Comment (RobGeada) [inline src/trustyai_service/service/data/datasources/data_source.py:283]: Unmoderated exception back to user

**Review by RobGeada — CHANGES_REQUESTED** [review]

**Comment by SudipSinha** [comment]
✅ **Addressed review feedback and rebased**

Fixed Rob's security concerns:
- **metadata.py:751** - Sanitized error message to hide model_id from users
- **data_source.py:272** - Removed unmoderated exception details from error message

Both errors now show generic "Check server logs for details" while full context goes to structured logs.

Also rebased onto `upstream/main` (now at e32cec6).

**Comment by SudipSinha** [comment]
✅ **Tests passing, linters clean**

All tests verified:
- 17/17 tag management tests pass
- 41/41 info endpoint tests pass  
- 109/109 data source tests pass

Linters clean (ruff, pyrefly).

File: src/trustyai_service/endpoints/metadata.py (hunk 3 above)
Comment (SudipSinha) [reply src/trustyai_service/endpoints/metadata.py:845]: Addressed in d12af1a512240006aad7c4af2af355ff66a747b1

File: src/trustyai_service/service/data/datasources/data_source.py (hunk 4 above)
Comment (SudipSinha) [reply src/trustyai_service/service/data/datasources/data_source.py:283]: Addressed in d12af1a512240006aad7c4af2af355ff66a747b1

**Review by RobGeada — APPROVED** [review]
